Create user profile

daisakuhazui opened the pull request:
# WHAT
ユーザープロフィールの閲覧・更新の機能を実装

# WHY
Userモデルがその他のモデルの親モデルになっているため、川上から機能を実装していきます。

Commit:
コントローラの削除

Railsの作法で、questions_controller.rbとすべきところをquestion_controller.rbとしていました。
ルーティングを追加している最中に気づき、question_controller.rbのファイル名をquestions_controller.rbに変更しました。

Commit:
userコントローラの作成

show, edit, updateアクションを定義しました。
各アクションごとにユーザーのidをparams[:id]で取得しています。

updateアクションではupdate_attributesメソッドで複数のカラムを更新できるようにしてます。また、不正なパラメータが入らないようストロングパラメータを利用しています。

Commit:
メニューの項目とリンク先を調整

Uupdate profileでnameやintroduceなどの編集画面に、Settingでemailやpasswordといったログイン情報の編集画面に飛べるようにしました。
（SNSを利用するユーザーは自分のプロフィールを頻繁に編集しますが、ログイン情報はほとんど触らないため）

questions_pathをquestions_pathに直したのはコントローラ名を変更したためです。

Commit:
ユーザープロフィールの編集画面を作成

emailとpassword以外のユーザー情報を編集できるページです。
genresの選択機能は後のブランチで実装します。

Commit:
ユーザープロフィールの閲覧画面を作成

genresと同じく、questionsまわりの機能を実装後にユーザーが投稿した質問も表示されるようにします。

Commit:
ルーティングの調整

コントローラ名を変更したため、question#indexではなくquestions#indexに修正しました。
resourcesメソッドを用いて、show, update, editのルーティングを行いました。

Commit:
フォルダ名の修正

question_controller.rbからquestions_controller.rbに名前を変更したため、フォルダ名もquestionからquestionsに変更しました。

ArakawaShintaro commented on app/controllers/questions_controller.rb:
```diff
@@ -0,0 +1,4 @@
+class QuestionsController < ApplicationController
```
ほんとはこの修正は別PRでやるべきですが、今回はこれで大丈夫です

daisakuhazui replied:
承知しました！

ArakawaShintaro commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,24 @@
+  end
+
+  def edit
+    @user = User.find(params[:id])
```
show, update, editで同じコードを書いてるのでここはDRYにしたいですね。

```
def set_user
  @user = User.find(params[:id])
end
```
みたいな感じでbefore_actionを使って呼び出しましょう

daisakuhazui replied:
了解です。ありがとうございます！

ArakawaShintaro commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,24 @@
+    if @user.update_attributes(user_params)
+      # この後、updateが成功したらflashメッセージが出るようにする
+    else
+      render 'edit'
```
エラー内容も通知されたら、もっといいですね

daisakuhazui replied:
その方が親切ですよね。了解です！

ArakawaShintaro commented on app/views/users/edit.html.erb:
```diff
@@ -0,0 +1,26 @@
+<%= provide(:tiele,"Update your profile") %>
```
```
(:tiele, "Update your profile")
```
半角スペースほしいです

daisakuhazui replied:
・・・見逃してました。直します。

ArakawaShintaro reviewed (commented):
ひとまず見ました！

Commit:
スペースの調整

Commit:
before_actionの設定

各アクションにあった同じコードを1箇所にまとめました。

ArakawaShintaro commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,27 @@
+private
+
+    def user_params
+      params.require(:user).permit(:email, :alias_name, :name, :profile_image, :cover_image, :area, :introduce )
```
最後の半角スペースが気になりますね

ArakawaShintaro replied:
あと1行80文字くらいが見やすいので、改行をよしなにしておきましょう

```
 params.require(:user).permit(
                                     :email, 
                                      :alias_name, 
                                      :name, 
                                      :profile_image, 
                                      :cover_image, 
                                      :area, 
                                        :introduce 
                                      )
```

ArakawaShintaro commented:
別PR gem2個追加

- rubocop
- pry-rails

ArakawaShintaro commented:
別PR
Userの画像周り

ArakawaShintaro commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,27 @@
+    end
+  end
+
+  def edit
```
edit update

Commit:
見やすいようにコードをリファクタリング

editアクションの記述位置を見直しました。
また、user_params内の表記を見やすいよう改行を加えました。

Commit:
pry-railsとrubocopのインストール

開発効率を上げるためにgemを導入しました。

Commit:
Userプロフィールの編集画面作成

8行目にてエラーメッセージが表示されるよう記述しました。

cover_imageとprofile_imageは登録済みの画像が表示された上で、新たに変更したいファイルをアップロードできるようにしてあります。また、uploaderにて保存時にリサイズをかけているため画像の表示サイズは指定していません。

Commit:
エラーメッセージを表示するためのパーシャル

Commit:
フラッシュメッセージの表示位置

ヘッダー上部にフラッシュメッセージが表示されるようにしました。

ArakawaShintaro commented on Gemfile:
```diff
@@ -5,51 +5,32 @@ git_source(:github) do |repo_name|
-  # Adds support for Capybara system testing and selenium driver
   gem 'capybara', '~> 2.13'
   gem 'selenium-webdriver'
+  gem 'pry-rails'
```
これ自体も本来は別PRでやるべきですね

Commit:
profile_imageのアップローダー作成

画像を保存する際、200px✕200px上限にリサイズしています。

Commit:
cover_imageのアップローダー作成

cover_imageの場合、一旦800px✕800pxにリサイズするようにしました。

Commit:
uploaderをUserモデルにマウント

ArakawaShintaro commented on app/uploaders/profile_uploader.rb:
```diff
@@ -0,0 +1,13 @@
+class ProfileUploader < CarrierWave::Uploader::Base
+  include CarrierWave::RMagick
+  process :resize_to_limit => [200, 200]
```
ハッシュロケットを使うのはやめましょう

```
resize_to_limit:  [200, 200]
```

ArakawaShintaro replied:
他にも色々そうなってるので全て修正してください

Commit:
シンボルへの書き換え

ハッシュロケットをシンボルへ書き直しました。

ArakawaShintaro reviewed (approved):
PRの僕のコメントには1つずつ返信いただいた方が助かります。
コメントを見ているか不安なので..
ひとまずマージで！
